File: app/src/main/java/dev/yiray/qickmvivm/base/SideEffectEvent.java
package dev.yiray.qickmvivm.base;

import android.util.Pair;

public class SideEffectEvent<S> {
    private Boolean hasBeenHandled = false;
    private Pair<S, S> sideEffectPair;

    public SideEffectEvent(S state, S sideEffect) {
        sideEffectPair = new Pair<>(state, sideEffect);
    }

    public Pair<S, S> getSideEffectPair() {
        if (hasBeenHandled) {
            return null;
        }

        hasBeenHandled = true;

        return sideEffectPair;
    }
}

File: app/src/main/java/dev/yiray/livedatamvi/base/SideEffectEvent.kt
package dev.yiray.livedatamvi.base

import android.util.Pair

class SideEffectEvent<S>(state: S, sideEffect: S) {
    private var hasBeenHandled = false
    val sideEffectPair: Pair<S, S> = Pair(state, sideEffect)

    fun getSideEffect(): Pair<S, S>? {
        if (hasBeenHandled) {
            return null
        }
        hasBeenHandled = true
        return sideEffectPair
    }
}